### ISSUE TITLE
Duration missing features

### ISSUE THREAD

**Issue by llacroix** [moment/moment#463]
I'm working on a project and momentjs is really useful for date manipulation so thanks for that. 

To give you an example, we're making somekind of store for plane tickets. The `duration.humanize()` function is way to imprecise for us. 

For example when leaving at:  12:00 and landing at 13:30. It will round it to 2 hours. But we need some kind of granularity which is missing from momentjs.

humanize could default to 1 level of precision starting from the highest value. 

    // 1:30
    duration.humanize()
    2 hour
    duration.humanize({precision: 2})
    1 hour 30 minutes
    duration.humanize({precision: 3})
    1 hour 30 minutes 0 seconds

The second problem is when there is the number 0. Momentjs translates it as "few seconds". I'm not so sure how it could be achieved but it would be cool to have some kind of formatting. 

    // 3 days and 0 minutes
    duration.format('D [and] M')
    > 3 days and 0 minutes
    duration.format('H [and] M')
    > 72 hours and 0 minutes
    duration.format('H [and] m')
    > 72 hours and few minutes

That way the highest value could be mapped in the format. So even if there is 1 year, the format tells us how to display it correctly. I'd be happy to push a commit for that because it's very useful and having to do it by hand when momentjs already handle localization feels bad.

**Comment by n1ghtmare** [comment]
:+1:

**Comment by hdias2310** [comment]
+1 duration.format!

**Comment by togakangaroo** [comment]
sooo....wat's the deal with this?

**Comment by icambron** [comment]
Just dropping in to note a few things, since there are like a million comments here. I'm not the maintainer and am no longer active in this project, but I'm pretty sure of two things:

 1. Everyone, you can just use @jsmreese's plugin. It looks great! Plugins are good things! (@jsmreese, if you're so inclined, you can add it to the list of Moment plugins in the documentation with a PR here: https://github.com/moment/momentjs.com)

 2. It's not going to be included in Moment's core until someone a) internationalizes it, b) integrates the tests into Moment's tests, and c) submits a PR. It will **definitely not happen until someone does that**.

**Comment by crush83** [comment]
+1 duration.format

**Comment by Iorg** [comment]
+1 duration.format!

**Comment by humpedli** [comment]
+1 duration.format

**Comment by LePhil** [comment]
+1 duration.format. Our case is that we want to provide a function where one can pass an amount of seconds and a string that would format them. Right now we're doing it with moment.startof("w").add("s", nrOfSeconds).format( customFormatString );

Unfortunately just last week this produced an error since startof("w") is Sunday and on Sunday it was time for daylight saving time, so we were missing an hour there. Of course we can check that with isDSTShifted() but it would still be awesome to not have to convert a duration to a date just to format it.

**Comment by CharlesOkwuagwu** [comment]
+1 duration.format

**Comment by CharlesOkwuagwu** [comment]
@jamesrom 

I'm trying to integrate your formatting with http://ionden.com/a/plugins/ion.rangeSlider/demo_advanced.html 

No JS errors ... but stops the slider from working, some assistance please

**Comment by chuanxd** [comment]
+1 duration.format

**Comment by mxpv** [comment]
+1 for ``duration.format``

**Comment by SamFromDaUk** [comment]
+1 for duration.format

**Comment by togakangaroo** [comment]
Can we close this as apparently nobody +1ing actually reads this thread?

**Comment by gkatsanos** [comment]
OR, fix it?

**Comment by bguiz** [comment]
George, I think those who are adding +1 s are implicitly disagreeing, and
simply want these features.

If the notifications are irritating you, consider unsubscribing from this
thread.

**Comment by togakangaroo** [comment]
The entire reason github doesn't have a vote feature (from my understanding) is because it discourages discussion.

You can implicitly disagree all you want, that's not helpful. If you disagree that's great, read through the thread and give your arguments (specifically, any new arguments or any points you think have been poorly considered).

I'm still subscribed because I *do* think there's some interesting arguments left to be made, I look forward to hearing them, but the team shouldn't bow to mere peer pressure.

**Comment by cmawhorter** [comment]
I can confirm the plugin @icambron mentioned above works.  

If you're like me, you want a `.fromNow()` with broader output (like amazon's shipping countdown).  Here's how to do that:

```javascript
// include the plugin
var cutoffDate = moment().add(3, 'hours'); // some date in the future that we're counting down to
var now = moment();
var output = moment.duration(cutoffDate.diff(now)).format('h [hours] m [minutes] s [seconds]');
// "3 hours 0 minutes 0 seconds"
```

**Comment by KristerV** [comment]
yep, I shall also make it clear - moment.duration().format() works with this plugin: https://github.com/jsmreese/moment-duration-format

if you're using meteor, go for `oaf:moment-duration-format`

**Comment by Markj89** [comment]
This doesn't seem to be working for me, I'm using moment .diff between two times, which does work, the problem is when It is between "pm" and "am", say I'm traveling at 9pm and landing 2am, I should get the correct amount of time.

**Comment by KristerV** [comment]
@Markj89 you need to open a stackoverflow topic.

**Comment by JobaDiniz** [comment]
+1

**Comment by dotnetwise** [comment]
+1

**Comment by aromanyukvakoms** [comment]
+1

**Comment by BlueHotDog** [comment]
:+1:

**Comment by zerocooljs** [comment]
+1

**Comment by gen4sp** [comment]
+1

**Comment by remypar5** [comment]
+1 on `.format()`

**Comment by inf3rno** [comment]
+1

**Comment by Hugheth** [comment]
+1

**Comment by inf3rno** [comment]
Btw. I think it should be possible to parse the humanized format as well. E.g. it is not that hard to parse `7 days and 5 hours`...

**Comment by abepetrillo** [comment]
Just wanted to emphasize @jsmreese's [plugin](https://github.com/jsmreese/moment-duration-format) worked great!

**Comment by sutkovoy** [comment]
:+1:

**Comment by qieqing** [comment]
+1 duration.format

**Comment by prophe05** [comment]
+1 duration format

**Comment by aboalwi** [comment]
+1 duration.format

**Comment by dangoosby** [comment]
+1 duration.format

**Comment by Asido** [comment]
:+1:

**Comment by maggiepint** [comment]
I'm going to close this in favor of #1048. It doesn't mean it's gone, we just had two issues that are several years old and both boil down to 'make duration formatting'. The other one explains this more clearly.

**Comment by wsain001** [comment]
+1 duration.format

**Comment by gbaumgart** [comment]
+1 duration.format

**Comment by ondrek** [comment]
+1

**Comment by rawthriver** [comment]
Use .localeData().relativeTime(45, true, 'mm') and get 45 minutes, and so on ;)

**Comment by vedmant** [comment]
+1 Really so much talk about this and yet it's not implemented. There is even separate package for this https://github.com/jsmreese/moment-duration-format

**Comment by elasticsteve** [comment]
I can't believe duration.format doesn't exist! Any particular reason? Am I missing something? I was searching for hours how to display a duration with minutes and seconds (mm:ss), but couldn't make it work. Does this not exist?

**Comment by sagivo** [comment]
to display duration in format, simply convert duration to ms and then to moment:
`moment(duration.as('milliseconds')).format('hh:mm:ss')`

**Comment by elasticsteve** [comment]
Thanks! Is this in the docs? I didn't see it...

**Comment by maggiepint** [comment]
This is not in the docs because it doesn't work. It is a hack using the unix epoch to interpret milliseconds as a point in time since the epoch. Thus, it falls apart if your duration is 24 hours or greater. 

A few things about this code:

1. An hh token is a token meant to be used with AM/PM, so your duration will only go up to 12

2. Even if you use an HH token, which will go up to 23, your duration cannot be any longer than 23:59:59.999, as things will cross over into the next date

3. As this code is written, it will only work on a server that is in UTC. Any browser in a different timezone will yield unexpected results.

If you want to use this hack it should be written as:

```js
moment.utc(duration.as('milliseconds')).format('HH:mm:ss')
```

But as I pointed out, this will only work if your duration is under 24 hours. Over that, back to square 1.

**Comment by elasticsteve** [comment]
Thanks. It took me a while today to figure out that one can concatenate `.minutes()` with `.seconds()`. Call me stupid, but it wasn't obvious from me in the docs what the difference between `.seconds()` and `.asSeconds()` is right away... But now I see it's actually there! :)
 
for example: `console.log(duration.minutes()+":"+duration.seconds());`

I'll stick with this method.

**Comment by butterflyhug** [comment]
@elasticsteve You'll probably also want to add a leading zero to the seconds portion if it's less than ten.  :wink:

**Comment by sagivo** [comment]
@butterflyhug @elasticsteve that's actually the main reason why i wanted to use format. in my case it's a simple short time counter so @maggiepint's solution is enough for my needs. 
`moment.utc(duration.asMilliseconds())).format('HH:mm:ss')`

**Comment by elasticsteve** [comment]
@sagivo  good point! Now the @maggiepint solution looks more appealing. My server is UTC anyway.

**Comment by butterflyhug** [comment]
Sure, I totally get it.  I've also used that hack in production code.  But I wouldn't recommend it as a general solution (apart from a few specific use cases), because there are lots of situations where it fails badly.  (E.g. negative durations are another important failure case that hasn't been explicitly mentioned yet.)

**Comment by qi-yu** [comment]
+1

**Comment by SlancikM** [comment]
+1 duration.format() !!!!!!!!!

**Comment by sebastianhaas** [comment]
+1

**Comment by mazing** [comment]
+1

**Comment by arbaaz** [comment]
+1

**Comment by stzonis** [comment]
+1

**Comment by BenDevelopment** [comment]
+1

**Comment by t4deu** [comment]
+1

**Comment by toneil** [comment]
+1

**Comment by serezhka** [comment]
+1

**Comment by denova** [comment]
+1

**Comment by ohtangza** [comment]
+1

**Comment by Maxpain177** [comment]
+1

**Comment by ffocsi** [comment]
+1

**Comment by FrozenDroid** [comment]
+1

**Comment by wallthis** [comment]
+1

**Comment by denl95** [comment]
+1

**Comment by aguynamedben** [comment]
+1 hard to debug durations

**Comment by zaplachinsky** [comment]
+1

**Comment by Kovbo** [comment]
You can easily do something like:

      ` var duration = moment(yourFutureValueHere) -  moment();
        var s = Math.floor( (duration/1000) % 60 );
        var m = Math.floor( (duration/1000/60) % 60 );
        var h = Math.floor( (duration/(1000*60*60)) % 24 );
        var d = Math.floor( duration/(1000*60*60*24) );
        return d + ' days ' + h + ' hours ' + m + ' minutes ' + s  + ' seconds ';   // 12 days 12 hours 12 minutes 12 seconds
`

**Comment by ToonLadyboy** [comment]
Request one user twitter @Toon_Ladyboy

**Comment by avnovoselov** [comment]
https://github.com/jsmreese/moment-duration-format

**Comment by alfalfarias** [comment]
> You can easily do something like:
> 
> ```
>     var duration = moment().add(5, 'days') -  moment().local();
>     var s = Math.floor( (duration/1000) % 60 );
>     var m = Math.floor( (duration/1000/60) % 60 );
>     var h = Math.floor( (duration/(1000*60*60)) % 24 );
>     var d = Math.floor( duration/(1000*60*60*24) );
>     return d + ' days ' + h + ' hours ' + m + ' minutes ' + s  + ' seconds before the deadline.';   
>     // 4 days 23 hours 59 minutes 59 seconds before the deadline.
> ```

Uggly, but effective.

**Comment by brennancheung** [comment]
I needed it in `HH:mm:ss`.  This is how I went about it.

`moment().startOf('year').add(theDuration).format('HH:mm:ss')`

**Comment by gkatsanos** [comment]
A simple parameter in humanize, .e.g `humanize(precise: true)` , which would circumvent all the rounding would be enough to make everyone happy. Besides, rounding is the hard part to implement. The rest is just simple converting a duration like any other formatting done with format().

**Comment by drydenwilliams** [comment]
There is a great `moment` method called `fromNow()` that will return the time from a specific time in nice human readable form, like this: 

```
moment('2019-04-30T07:30:53.000Z').fromNow() // an hour ago || a day ago || 10 days ago
```

Or if you want that between two specific dates you can use: 

```
var a = moment([2007, 0, 28]);
var b = moment([2007, 0, 29]);
a.from(b); // "a day ago"
```

Taken from the Docs: 

 - [`fromNow()`](https://momentjs.com/docs/#/displaying/fromnow/)
 - [`from()`](https://momentjs.com/docs/#/displaying/from/)

**Comment by louwers** [comment]
Can this be reopened? This is very basic functionality.

**Comment by schnetzi** [comment]
For my case I found a package which solved my problems:
https://github.com/EvanHahn/HumanizeDuration.js

Maybe it is useful for someone else as well. :)

**Comment by elhakeem** [comment]
July 2019 and this feature still not available.

**Comment by kawaaii** [comment]
> July 2019 and this feature still not available.

Haven't tried it yet, but I just noticed this!
`npm install moment-duration-format`

https://github.com/jsmreese/moment-duration-format

**Comment by Thaina** [comment]
September 2019 still officially miss this feature

```js
d.format('H:mm:ss');
"1:01:01"
```

Also wish that it could have this feature

```js
d.format('D-H:mm:ss'); // existence of D will mod hours to 24
"999-23:59:59"
```

**Comment by laggingreflex** [comment]
In case anyone was wondering why this was closed: https://github.com/moment/moment/issues/463#issuecomment-228543000

**Comment by kmjungersen** [comment]
This is by no means ideal, but I ended up doing something like this as a workaround.  I had a `duration` object with a countdown for logging the user out after inactivity, and wanted to format the time remaining. 

```typescript
// What I did:
private getTimeString(duration: moment.Duration): string {
    const time = moment()
      .seconds(duration.seconds())
      .minutes(duration.minutes());

    return time.format('mm:ss');
}

// What I'd rather do (as many others have mentioned)...
private getTimeString(duration: moment.Duration): string {
    return duration.format('mm:ss');
}
```

**Comment by jimmy-e** [comment]
A hacky solution of mine:

```js
import moment from 'moment';

const formatInt = (int: number): string => {
  if (int < 10) {
    return `0${int}`;
  }
  return `${int}`;
};

export const formatDuration = (time: string): string => {
  const seconds = moment.duration(time).seconds();
  const minutes = moment.duration(time).minutes();
  const hours = moment.duration(time).hours();
  if (hours > 0) {
    return `${formatInt(hours)}:${formatInt(minutes)}:${formatInt(seconds)}`;
  }
  if (minutes > 0) {
    return `${formatInt(minutes)}:${formatInt(seconds)}`;
  }
  return `00:${formatInt(seconds)}`;
};

```

**Comment by justrealmilk** [comment]
2020

**Comment by vjfuenzalida** [comment]
I use the following method, it may be useful to someone else :) 

```javascript
function formatDuration(duration, format) {
  const date = moment().startOf('day');
  return date.add(duration).format(format);
}
```
(I use only to format durations between 00:00 and 23:59)

**Comment by GuitooStephan** [comment]
Still need this

**Comment by dazoliveira** [comment]
These 'get' helps to format in way u want:

      duration.get('years')
      duration.get('months')
      duration.get('days')
      duration.get('hours')
      duration.get('minutes')
      duration.get('seconds')

**Comment by elisherer** [comment]
What I work with:
```js
const formatDuration = ms => {
  const days = Math.floor(ms / 8.64e7);
  const msOnLastDay = ms - days * 8.64e7;
  return (days < 10 ? "0" + days : days) + ":" + moment.utc(msOnLastDay).format("HH:mm:ss.SSS");
};
```

```js
formatDuration(5)
"00:00:00:00.005"
formatDuration(500)
"00:00:00:00.500"
formatDuration(50000)
"00:00:00:50.000"
formatDuration(5000000)
"00:01:23:20.000"
formatDuration(500000000)
"05:18:53:20.000"
// for reference
JSON.stringify(moment.duration(500000000)._data, null, 2)
"{
  "milliseconds": 0,
  "seconds": 20,
  "minutes": 53,
  "hours": 18,
  "days": 5,
  "months": 0,
  "years": 0
}"
```

**Comment by VADS** [comment]
> > July 2019 and this feature still not available.
> 
> Haven't tried it yet, but I just noticed this!
> `npm install moment-duration-format`
> 
> https://github.com/jsmreese/moment-duration-format

This works perfectly fine :)

**Comment by Muskos** [comment]
The game is over :)

**Comment by dazoliveira** [comment]
Thanks, appreciate that!!!
________________________________
De: Aleksey Makas <notifications@github.com>
Enviado: sexta-feira, 9 de outubro de 2020 11:03
Para: moment/moment <moment@noreply.github.com>
Cc: Douglas Aguiar <douglaszanquetta@hotmail.com>; Comment <comment@noreply.github.com>
Assunto: Re: [moment/moment] Duration missing features (#463)


The game is over :)

**Comment by ilhamarrouf** [comment]
2021, still in need of this feature.

**Comment by qa-sim** [comment]
2022, still in need of this.

**Comment by Phecda** [comment]
My solution for a simple `HH:mm:ss` is
``` typescript
export function formatDuration(duration: moment.Duration) {
  // convert 'P1D' to 'PT24H'
  const isoString = moment.duration(duration.asMilliseconds()).toISOString();

  // const [, , h = '', , m = '', , s = '', , ms = ''] =
  const [, , h = '', , m = '', , s = ''] =
    isoString.match(/T((\d+)H)?((\d+)M)?(([\d]+)(\.(\d+))?S)?/) ?? [];

  // return `${h.padStart(2, '0')}:${m.padStart(2, '0')}:${s.padStart(2, '0')}.${ms.padStart(3, '0')}`;
  return `${h.padStart(2, '0')}:${m.padStart(2, '0')}:${s.padStart(2, '0')}`;
}
```

see https://momentjs.com/docs/#/durations/as-iso-string/

**Comment by TravnikovDev** [comment]
Summer 2022, we still in need of this feature...

**Comment by Yuniac** [comment]
I'm surprised people have been asking for this feature for years and it's still not supported lol.

**Comment by schmod** [comment]
As has been repeatedly discussed in this thread, internationalization concerns make it difficult-to-impossible to implement this feature with any sort of reasonable API.

If you'd like to implement a library that implements this feature for a single locale, you're free to do so (there are numerous examples of this referenced above). However, working around the grammars of every locale that Moment supports (and gathering the relevant translation data) is a much more difficult proposition.  

If you've figured out how to build an i18n-friendly version of this API, you're also free to build that yourself.  (Note that you'll need to fork Moment, as the project [is in "maintenance mode"](https://momentjs.com/docs/#/-project-status/), and the maintainers are not considering major new features.)

Given Moment's deprecated status, you may want to consider making this a proper TC39 [proposal](https://github.com/tc39/ecma262/blob/HEAD/CONTRIBUTING.md) as an extension of the [Temporal API](https://github.com/tc39/proposal-temporal).  I'm sure that many would consider this to be a valuable addition to the ECMAScript standard library.

**Comment by inf3rno** [comment]
@schmod Partial support exists, but if you never start to implement it because it is too hard to do full support, then you will never complete it. I guess there are many people here, who know different languages, so adding a language would be just starting an enchancement issue and sumbitting some simple code.

**Comment by foo-bar-dr** [comment]
Summer 2023, still in need of this feature.
